core/drivers/clk/clk_dt.c

```diff
@@ -0,0 +1,395 @@
+		if (drv->type != DT_DRIVER_CLK)
+			continue;
+
+		clk_drv = (const struct clk_driver *)drv->driver;
+		for (dm = drv->match_table; dm && dm->compatible; dm++) {
+			if (strcmp(dm->compatible, compat) == 0)
+				return clk_drv;
+		}
+	}
+
+	return NULL;
+}
+
+/* Recursively called from parse_clock_property() */
+static TEE_Result clk_probe_clock_provider_node(const void *fdt, int node);
+
+static TEE_Result parse_clock_property(const void *fdt, int node)
+{
+	int len = 0;
+	int idx = 0;
+	int parent_node = 0;
+	int clock_cells = 0;
+	uint32_t phandle = 0;
+	const uint32_t *prop = NULL;
+	TEE_Result res = TEE_ERROR_GENERIC;
+
+	prop = fdt_getprop(fdt, node, "clocks", &len);
+	if (!prop)
+		return TEE_SUCCESS;
+
+	len /= sizeof(uint32_t);
+	while (idx < len) {
+		phandle = fdt32_to_cpu(prop[idx]);
+
+		parent_node = fdt_node_offset_by_phandle(fdt, phandle);
+		if (parent_node < 0)
+			return TEE_ERROR_GENERIC;
+
+		/* Parent setup should not fail or clock won't be available */
+		res = clk_probe_clock_provider_node(fdt, parent_node);
+		if (res)
+			panic("Failed to setup parent clock");
+
+		clock_cells = fdt_clock_cells(fdt, parent_node);
+		idx += 1 + clock_cells;
+	}
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result clk_dt_node_clock_setup_driver(const void *fdt, int node)
+{
+	int idx = 0;
+	int len = 0;
+	int count = 0;
+	const char *compat = NULL;
+	TEE_Result res = TEE_ERROR_GENERIC;
+	const struct clk_driver *clk_drv = NULL;
+
+	count = fdt_stringlist_count(fdt, node, "compatible");
+	if (count < 0)
+		return TEE_ERROR_GENERIC;
+
+	for (idx = 0; idx < count; idx++) {
+		compat = fdt_stringlist_get(fdt, node, "compatible", idx,
+					    &len);
+		if (!compat)
+			return TEE_ERROR_GENERIC;
+
+		clk_drv = clk_get_compatible_driver(compat);
+		if (!clk_drv)
+			continue;
+
+		res = clk_drv->setup(fdt, node);
+		if (res != TEE_SUCCESS) {
+			EMSG("Failed to setup clock driver for compatible %s",
+			     compat);
+			panic();
+		} else {
+			return TEE_SUCCESS;
+		}
+	}
+
+	return TEE_ERROR_GENERIC;
+}
+
+static void clk_probe_node(const void *fdt, int parent_node);
+
+static TEE_Result clk_probe_clock_provider_node(const void *fdt, int node)
+{
+	int len = 0;
+	int status = 0;
+	TEE_Result res = TEE_ERROR_GENERIC;
+
+	status = _fdt_get_status(fdt, node);
+	if (!(status & DT_STATUS_OK_SEC))
+		return TEE_ERROR_ITEM_NOT_FOUND;
+
+	/* Check if the node is a clock provider */
+	if (!fdt_getprop(fdt, node, "#clock-cells", &len))
+		return TEE_ERROR_ITEM_NOT_FOUND;
+
+	/* Check if node has already been probed */
+	if (clk_get_provider_by_node(node))
+		return TEE_SUCCESS;
+
+	/* Check if the node has a clock property first to setup parent */
+	res = parse_clock_property(fdt, node);
+	if (res)
+		return res;
+
+	return clk_dt_node_clock_setup_driver(fdt, node);
+}
+
+static void clk_probe_node(const void *fdt, int parent_node)
+{
+	int child = 0;
+
+	fdt_for_each_subnode(child, fdt, parent_node) {
+		clk_probe_clock_provider_node(fdt, child);
```

Comment:
`clk_probe_clock_provider_node()` return code should checked, at least asserted against unexpected values:
```c
		res = clk_probe_clock_provider_node(fdt, child);
		assert(res == TEE_SUCCESS || res == TEE_ERROR_ITEM_NOT_FOUND);

Reply:
Indeed